### PR TITLE
profiler for powdr asm compiled from rust

### PR DESCRIPTION
**Pull request by pacheco**
Rust profiler (#1174) implemented by tracking the execution of the `riscv-executor`.

### CODE REVIEW

File: riscv-executor/src/lib.rs [hunk 1]
```diff
@@ -1073,6 +1138,7 @@ pub fn execute<F: FieldElement>(
     inputs: &Callback<F>,
     bootloader_inputs: &[Elem<F>],
     mode: ExecMode,
+    profiler_opt: ProfilerOptions,
```
Comment (leonardoalt) [inline riscv-executor/src/lib.rs:1141]: Should this maybe be an Option?

File: riscv-executor/src/lib.rs [hunk 2]
```diff
@@ -979,20 +1003,33 @@ impl<'a, 'b, F: FieldElement> Executor<'a, 'b, F> {
     }
 }
 
+/// return true if the expression is a jump instruction
+fn expr_is_jump(e: &Expression) -> bool {
```
Comment (leonardoalt) [inline riscv-executor/src/lib.rs:1007]:
```suggestion
fn is_jump(e: &Expression) -> bool {
```
arg is already Expr

File: cli-rs/src/main.rs [hunk 3]
```diff
@@ -122,6 +124,16 @@ enum Commands {
         #[arg(short, long)]
         #[arg(default_value_t = false)]
         witness: bool,
+
+        /// Generate a flamegraph plot of the execution ("[file].svg")
+        #[arg(long)]
+        #[arg(default_value_t = false)]
+        generate_flamegraph: bool,
```
Comment (leonardoalt) [inline cli-rs/src/main.rs:131]:
```suggestion
        flamegraph: bool,
```
maybe simpler?

File: cli-rs/src/main.rs [hunk 4]
```diff
@@ -122,6 +124,16 @@ enum Commands {
...
+
+        /// Generate callgrind file of the execution ("[file].callgrind")
+        #[arg(long)]
+        #[arg(default_value_t = false)]
+        generate_callgrind: bool,
```
Comment (leonardoalt) [inline cli-rs/src/main.rs:136]:
```suggestion
        callgrind: bool,
```
maybe simpler?

File: riscv-executor/src/profiler.rs [hunk 5]
```diff
@@ -0,0 +1,346 @@
...
+
+use rustc_demangle::demangle;
+
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub struct Call<'a> {
+    from: Loc<'a>,
+    target: Loc<'a>,
+}
+
+/// Risc-v asm profiler.
```
Comment (leonardoalt) [inline riscv-executor/src/profiler.rs:19]:
```suggestion
/// RISCV asm profiler.
```

File: riscv-executor/src/profiler.rs [hunk 6]
```diff
@@ -0,0 +1,346 @@
...
+/// Tracks the self-cost of functions and cummulative cost specific function calls (i.e., callgrind style).
```
Comment (leonardoalt) [inline riscv-executor/src/profiler.rs:20]:
```suggestion
/// Tracks the self-cost of functions and the cumulative cost of specific function calls (i.e., callgrind style).
```
?

File: riscv-executor/src/profiler.rs [hunk 7]
```diff
@@ -0,0 +1,346 @@
...
+    location_stats: BTreeMap<Loc<'a>, usize>,
+    /// (count, cummulative cost) of calls
+    call_stats: BTreeMap<Call<'a>, (usize, usize)>,
+    /// stack sampling format for FlameGraph
+    folded_stack_stats: BTreeMap<Vec<&'a str>, usize>,
+}
+
+#[derive(Default, Clone)]
+pub struct ProfilerOptions {
+    pub output_directory: String,
```
Comment (leonardoalt) [inline riscv-executor/src/profiler.rs:44]: In other crates we try to keep FS access to a minimal, and rather return results that CLI then writes to the FS. I guess we could do the same here without really disrupting the architecture?

Comment (leonardoalt) [reply riscv-executor/src/profiler.rs:44]: or you could take an `impl Write` here instead of the file path

File: riscv-executor/src/profiler.rs [hunk 8]
```diff
@@ -0,0 +1,346 @@
...
+        }
+    }
+
+    /// profiling only starts once "__runtime_start" is reached
+    pub fn running(&self) -> bool {
+        !self.call_stack.is_empty()
+    }
+
+    /// function at the top of the call stack
+    pub fn curr_function(&self) -> Option<&'a str> {
```
Comment (leonardoalt) [inline riscv-executor/src/profiler.rs:185]:
```suggestion
    pub fn current_function(&self) -> Option<&'a str> {
```

File: riscv-executor/src/profiler.rs [hunk 9]
```diff
@@ -0,0 +1,346 @@
...
+            self.write_flamegraph(&path);
+        }
+        if self.options.generate_callgrind {
+            path.set_extension("callgrind");
+            self.write_callgrind(&path);
+        }
+    }
+
+    /// profiling only starts once "__runtime_start" is reached
+    pub fn running(&self) -> bool {
```
Comment (leonardoalt) [inline riscv-executor/src/profiler.rs:180]:
```suggestion
    pub fn is_running(&self) -> bool {
```

File: riscv-executor/src/profiler.rs [hunk 10]
```diff
@@ -0,0 +1,346 @@
...
+            })
+            .collect();
+        let mut options = Default::default();
+        let file = File::create(path).unwrap();
+        let w = BufWriter::new(file);
+        inferno::flamegraph::from_lines(&mut options, lines.iter().map(|s| s.as_str()), w).unwrap();
+    }
+
+    /// calculate totals and write out results
+    pub fn execution_finished(&mut self) {
```
Comment (leonardoalt) [inline riscv-executor/src/profiler.rs:163]:
```suggestion
    pub fn finish(&mut self) {
```
maybe?

File: riscv-executor/src/profiler.rs [hunk 11]
```diff
@@ -0,0 +1,346 @@
...
+
+        if self
+            .return_pc_stack
+            .last()
+            .is_some_and(|saved_pc| *saved_pc == target_pc)
+        {
+            // "return" from current function
+            let (done_call, cost) = self.call_stack.pop().unwrap();
+            self.return_pc_stack.pop();
+            // add to calls cummulative cost and to parent call
```
Comment (leonardoalt) [inline riscv-executor/src/profiler.rs:306]:
```suggestion
            // add to calls cumulative cost and to parent call
```

Comment (leonardoalt) [reply riscv-executor/src/profiler.rs:306]: wdym by `to parent call`?

File: riscv-executor/src/profiler.rs [hunk 12]
```diff
@@ -0,0 +1,346 @@
...
+                *curr_cost += cost;
+            }
+        } else {
+            let target = self.location_at(target_pc).unwrap();
+            let curr_function = self.curr_function().unwrap();
+            if target.function != curr_function {
+                // "tail call": replace the current call in the stack
+                let (done_call, cost) = self.call_stack.pop().unwrap();
+
+                // add to calls cummulative cost and to parent call
```
Comment (leonardoalt) [inline riscv-executor/src/profiler.rs:318]: same here

File: riscv-executor/src/profiler.rs [hunk 13]
```diff
@@ -0,0 +1,346 @@
...
+            } else {
+                // "control flow" (or "tail call" to self, if that is a thing), don't think this needs special handling
+            }
+        }
+    }
+}
+
+fn format_function_name(name: &str) -> String {
+    if let Some(prefix) = name.find("___ZN") {
+        // format!("{}_{}", &name[0..4], demangle(&name[prefix + 2..]))
```
Comment (leonardoalt) [inline riscv-executor/src/profiler.rs:341]: what about this?

**Review by leonardoalt — COMMENTED** [review]
Nice! Mostly small comments

File: riscv-executor/src/profiler.rs (hunk 11 above)
Comment (pacheco) [reply riscv-executor/src/profiler.rs:306]: we update two costs here (e.g., lets say this is `bar` calling `foo` from line 20):
- cumulative cost of this specific call (`foo -> bar` at line 20)
- running cost of caller (`foo`)

**Review by leonardoalt — APPROVED** [review]

**Pull request merged**